lisa/tools/fio.py

```diff
@@ -0,0 +1,202 @@
+        self.node.execute(
+            "ln -s /usr/local/bin/fio /usr/bin/fio", sudo=True, cwd=code_path
+        ).assert_exit_code()
+        return self._check_exists()
+
+    def _get_iops_latency_by_qdepth(
+        self, qdepth: int, fio_results_list: List[FIOResult]
+    ) -> Dict[str, Decimal]:
+        mode_iops_latecny: Dict[str, Any] = {}
+        for fio_result in fio_results_list:
+            if qdepth == fio_result.qdepth:
+                mode_iops_latecny[f"{fio_result.mode}_iops"] = fio_result.iops
+                mode_iops_latecny[f"{fio_result.mode}_lat_usec"] = fio_result.latency
+                mode_iops_latecny["iodepth"] = fio_result.iodepth
+                mode_iops_latecny["qdepth"] = fio_result.qdepth
+        return mode_iops_latecny
+
+    def install(self) -> bool:
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages("fio")
+        if not self._check_exists():
+            self._install_from_src()
+        return self._check_exists()
+
+    def launch(
+        self,
+        name: str,
+        filename: str,
+        mode: str,
+        iodepth: int,
+        numjob: int,
+        time: int = 60,
+        block_size: str = "4K",
+        size_gb: int = 0,
+        direct: bool = True,
+        gtod_reduce: bool = True,
+        ioengine: str = "libaio",
+        group_reporting: bool = True,
+        overwrite: bool = False,
+    ) -> FIOResult:
+        cmd = (
+            f"--ioengine={ioengine} --bs={block_size} --filename={filename} "
+            f"--readwrite={mode} --runtime={time} --iodepth={iodepth} "
+            f"--numjob={numjob} --name={name}"
+        )
+        if direct:
+            cmd += " --direct=1"
+        if gtod_reduce:
+            cmd += " --gtod_reduce=1"
+        if size_gb:
+            cmd += f" --size={size_gb}G"
+        if group_reporting:
+            cmd += " --group_reporting"
+        if overwrite:
+            cmd += " --overwrite=1"
+        result = self.run(
+            cmd,
+            force_run=True,
+            sudo=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message=f"fail to run {cmd}",
+        )
+        matched_results = self._result_pattern.match(result.stdout)
+        assert matched_results
+        iops = matched_results.group("iops")
+        if iops.endswith("k"):
+            iops_value = Decimal(iops[:-1]) * 1000
+        else:
+            iops_value = Decimal(iops)
+        latency = matched_results.group("latency")
+        fio_result = FIOResult()
+        fio_result.iops = iops_value
+        fio_result.latency = Decimal(latency)
+        fio_result.iodepth = iodepth
+        fio_result.qdepth = iodepth * numjob
+        fio_result.mode = mode
+        return fio_result
+
+    def parse_test_results(
+        self, fio_results_list: List[FIOResult]
+    ) -> List[FIOResultsMessage]:
+        qdepth_list = sorted(set([x.qdepth for x in fio_results_list if x.qdepth]))
```

Comment:
can we use iodepth directly? the qdepth is for compatibility only.

Reply:
current dashboard is represent with qdepth , iodepth has duplicated values

Reply:
Let's use iodepth in code. The qdepth is used in current report, and we will use iodepth in new reports.

Reply:
the iodepths we used are not unique when run fio against ssd disks, not sure the root cause.

Reply:
What means "not unique"? If it's a problem, why isn't a problem when using qdepth?

Reply:
![image](https://user-images.githubusercontent.com/10083705/147338727-c4190da1-31da-4f8f-a5b8-331058d843f9.png)
they have duplicated values

Reply:
I see. But how the number of jobs is decided? It's the reason that the iodepth is not unique.

Reply:
I don't have the answer yet.

Reply:
I will try with number of cpu count when run against premium disk.

Reply:
all of them use core count as numjobs.